ejb3/src/main/java/org/jboss/as/ejb3/remote/AssociationImpl.java

```diff
@@ -73,23 +73,59 @@
 import java.util.Map;
 import java.util.Set;
 import java.util.concurrent.CancellationException;
+import java.util.concurrent.ConcurrentHashMap;
 import java.util.concurrent.Executor;
 import java.util.concurrent.Future;
 import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.stream.Collectors;
 
 /**
  * @author <a href="mailto:tadamski@redhat.com">Tomasz Adamski</a>
  */
-final class AssociationImpl implements Association {
+final class AssociationImpl implements Association, Registry.Listener<String, List<ClientMapping>>, AutoCloseable {
```

Comment:
"Why implement the listener interface directly?"
Because this object listens for registry events. 

"Are there functional changes hidden in here or is it just a code-shuffle refactor?"
In this class specifically?  The functional changes are:
* The Association object operates directly upon a specific Registry instead of one lazily supplied by the RegistryCollector.
* Replace Registry listener per connection with a single Registry listener for the Association object.  Registry events are queued and processed by a single thread per listener to maintain ordering, thus using a listener per connection is not appropriate.  The Association object maintains a separate collection of connection-specific ClusterTopologyListeners, to which Registry events are delegated.
* Ensure initial topology is set before attempting to process topology changes via Registry listener (see comments above the corresponding synchronization blocks).